### PR TITLE
Push some warnings for CSG non manifold and other errors.

### PR DESCRIPTION
**Pull request by fire**
@BastiaanOlij mentioned that seeing the error message for invalid CSGMesh inputs was complex.

This was also a leftover task from Fix mesh corruption of CSG by using #94321

Error Message Screenshot

I'm not perfectly happy with the warning wording. You are welcome to suggest changes.

Part of #100014

See also #100164

https://github.com/user-attachments/assets/f3080d1e

### CODE REVIEW

**Comment by fire** [comment]
Question, should child nodes error if their root id broken?

**Comment by Zireael07** [comment]
I say yes, to make it SUPER clear to user that things are broken

**Comment by fire** [comment]
@Zireael07 enhanced

**Comment by SpockBauru** [comment]
I don't think people will know what manifold means, maybe pointing a link to the CSG docs or just explaining that is not a "closed mesh" is better. "Manifold" is too cryptic.

**Comment by fire** [comment]
Thanks for the review! I'll try to put the entire definition in a short way.

**Comment by fire** [comment]
> "A manifold mesh is a 3D model that forms a solid object without any gaps, holes, or loose edges. The CSGShape3D could not be created because of this error: %s. For a mesh to be manifold, every edge must be shared by exactly two triangles, and the vertices of these triangles must be ordered correctly. The vertices should appear in a clockwise direction when viewed from the outside. This error has caused the CSGShape3D to be empty."

This is a draft error message. Will try it.

**Comment by SpockBauru** [comment]
Tested the windows artifact.

The warning does not appear when you set an invalid mesh, it only changes if there are changes in the tree. It also doesn't disappear when set a valid mesh.

Also when hovering the mouse the line in the in bottom is not splitting and goes over the screen limit of the monitor.


https://github.com/user-attachments/assets/c59f5ba1-ffe7-4c37-a6c7-4e54268c9d6c

**Comment by fire** [comment]
That's a different bug which has a pull request. https://github.com/godotengine/godot/pull/100340

**Comment by fire** [comment]
> Also when hovering the mouse the line in the in bottom is not splitting and goes over the screen limit of the monitor.

I wish I had some ideas with shorter wording, but I was stuck on the words. 

I feel like it's both my phrasing being too long and the tooltip not cutting the text.

**Comment by fire** [comment]
@SpockBauru Can you retest after the build is done?

**Comment by SpockBauru** [comment]
Tested the Windows artifact and there are no changes on my end, the hint box still goes over the screen limit.

image

File: modules/csg/csg_shape.cpp [hunk 1]
```diff
@@ -442,7 +434,42 @@ CSGBrush *CSGShape3D::_get_brush() {
...
+				child->warnings.clear();
+			}
+		}
+	}
+	const String manifold_error_message_template = "%s: The CSGShape3D could not be created due to this error, resulting in an empty shape.";
+	const String manifold_info_message = "CSGShape3D empty shapes typically occur because the mesh is not manifold. "
+										 "A manifold mesh forms a solid object without gaps, holes, or loose edges. "
+										 "For a mesh to be manifold, every edge of every triangle must contain the same two vertices (by index) as exactly one other triangle edge, "
+										 "and the start and end vertices must switch between these two edges. The Godot Engine triangle vertices must appear in clockwise order "
+										 "when viewed from the outside of the mesh.";
```
Comment (tetrapod00) [inline modules/csg/csg_shape.cpp:452]:
```suggestion
										 "A manifold mesh forms a solid object without gaps, holes, or loose edges. Each edge must connect to exactly two faces."
```
or
```suggestion
										 "A manifold mesh forms a solid object without gaps, holes, or loose edges. Each edge must be a member of exactly two faces."
```

I feel like this is about the right amount of information to put in a warning message: a one or two sentence summary, in friendly and relatively non-technical language.

Even as someone who understands all the concepts, the full-length message is very intimidating. My eyes start to glaze over as I read it. If we need a long and technical explanation like that, it should be in the docs somewhere, not in a warning popup. 

Maybe we can put it in the manual, along with some tips on how to troubleshoot non-manifold meshes.

Comment (fire) [reply modules/csg/csg_shape.cpp]: @tetrapod00 I have a draft of the process for making manifold shapes using Blender's fixing tool. https://docs.google.com/document/d/11KPDwMuZl3c4NSwanhatekAIvUxktURmXJVPkP_a17o/edit?tab=t.0#heading=h.r41iui30hwv7

**Comment by fire** [comment]
I lost the error by dragging the broken Suzanne node in the scene tree.

image

No idea why.

**Comment by SpockBauru** [comment]
Tested the windows artifact, the new wording seems to be good enough, it roughly explains the issue and the hint box does not goes over the screen limit anymore :D

Seems like #100340 has been merged. Is that applied to this PR? Because I'm still not triggering the error when selecting an invalid mesh. The error only triggers when there are changes in the tree, not in the mesh itself.

**Comment by fire** [comment]
Setting a mesh on csgmesh3d still is broken with respect to updating war ings

**Comment by fire** [comment]
I can now move the CSG nodes around and keep the warnings, but I lost the specific error message, and setting a mesh on a CSGMesh3D still does not update until the scene reloads.

File: modules/csg/csg_shape.cpp [hunk 2]
```diff
@@ -943,6 +901,23 @@ Array CSGShape3D::get_meshes() const {
...
+PackedStringArray CSGShape3D::get_configuration_warnings() const {
+	PackedStringArray warnings = Node::get_configuration_warnings();
+	const String manifold_error_message = "The CSGShape3D has an empty shape.";
+	const String manifold_info_message = "CSGShape3D empty shapes typically occur because the mesh is not manifold. "
+										 "A manifold mesh forms a solid object without gaps, holes, or loose edges. Each edge must be a member of exactly two faces.";
+	const CSGShape3D *current_shape = this;
+	while (current_shape) {
+		if (!current_shape->brush || current_shape->brush->faces.is_empty()) {
+			warnings.push_back(manifold_error_message);
+			warnings.push_back(manifold_info_message);
```
Comment (Calinou) [inline modules/csg/csg_shape.cpp:913]: Declare a single warning (otherwise, it's counted as two separate warnings by the editor) and make it localizable:

```suggestion
			warnings.push_back(RTR("The CSGShape3D has an empty shape.\nCSGShape3D empty shapes typically occur because the mesh is not manifold.\nA manifold mesh forms a solid object without gaps, holes, or loose edges.\nEach edge must be a member of exactly two faces."));
```

Comment (fire) [reply modules/csg/csg_shape.cpp]: > We should fix this in the dialog display instead.

I'm not sure if this is a bug or a feature enhancement. Can you write it up?

**Review by Calinou — APPROVED** [review]

**Comment by SpockBauru** [comment]
Tested the Windows artifact, the new tooltip is a way more readable:

image

**Comment by 0xafbf** [comment]
For the explanation of the two edges, I think it can be useful to refer to the normals of the faces for explaining the concept of manifold. I suggest something like this:
> Every edge of every triangle should link exactly two triangles, and the normals of the triangles should always face outwards
Although I don't know if all-inward faces are allowed.

**Comment by fire** [comment]
The nuance about all the orientations being the same winding order was removed due to being too technical. The 3mf manifold definition doesn’t require normals

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!
